File: content/realm/examples/node/legacy/Examples/client-reset.js
import Realm from "realm";
import { ObjectId } from "bson";

const REALM_APP_ID = "myapp-zufnj";

let realm;
let app;
beforeEach(async () => {
  app = new Realm.App({ id: REALM_APP_ID });
  await app.logIn(new Realm.Credentials.anonymous());
});
afterEach(async () => {
  if (realm && !realm.isClosed) {
    realm.write(() => {
      realm.deleteAll();
    });
    const path = realm.path;
    realm.close();
    Realm.deleteFile({ path });
    await app?.currentUser?.logOut();
  }
});
describe.skip("Client Reset with Seamless Loss", () => {
  // these tests can take longer than most, causing timeouts that make
  // the Github Actions CI fail
  jest.setTimeout(20000);

  const DogSchema = {
    name: "Doggo3",
    properties: {
      _id: "objectId",
      name: "string",
      age: "int?",
    },
    primaryKey: "_id",
  };
  test("Discard unsynced changes", async () => {
    let beforeCalled = false;
    let afterCalled = false;
    const clientResetSuccess = await new Promise((resolve, reject) => {
      // :snippet-start: discard-unsynced-changes
      const config = {
        schema: [DogSchema],
        sync: {
          user: app.currentUser,
          flexible: true,
          clientReset: {
            mode: "discardUnsyncedChanges",
            onBefore: (realm) => {
              console.log("Beginning client reset for ", realm.path);
              // :remove-start:
              beforeCalled = true;
              // :remove-end:
            },
            onAfter: (beforeRealm, afterRealm) => {
              console.log("Finished client reset for", beforeRealm.path);
              console.log("New realm path", afterRealm.path);
              // :remove-start:
              afterCalled = true;
              resolve(beforeCalled && afterCalled);
              // :remove-end:
            },
          },
        },
      };
      // :snippet-end:
      Realm.open(config).then((openedRealm) => {
        realm = openedRealm;
        realm.write(() => {
          realm.create("Doggo3", {
            _id: new ObjectId(),
            name: "Chippy",
            age: 12,
          });
          realm.create("Doggo3", {
            _id: new ObjectId(),
            name: "Jasper",
            age: 11,
          });
        });
        realm.syncSession._simulateError(
          211,
          "Simulate Client Reset",
          "realm::sync::ProtocolError",
          false
        );
      });
    });
    expect(clientResetSuccess).toBe(true);
  });

  test.skip("Recover Unsynced Changes Mode", async () => {
    // :snippet-start: recover-unsynced-changes
    const config = {
      schema: [DogSchema],
      sync: {
        user: app.currentUser,
        flexible: true,
        clientReset: {
          mode: "recoverUnsyncedChanges",
          onBefore: (realm) => {
            // This block could be used for custom recovery, reporting, debugging etc.
          },
          onAfter: (beforeRealm, afterRealm) => {
            // This block could be used for custom recovery, reporting, debugging etc.
          },
          onFallback: (session, path) => {
            // See below "Manual Client Reset Fallback" section for example
          },
        },
      },
    };
    // :snippet-end:
  });

  // TODO(DOCSP-23425): investigate real example once sdk changes merged
  test.skip("Recover or Discard Unsynced Changes Mode", async () => {
    // :snippet-start: recover-or-discard-unsynced-changes
    const config = {
      schema: [DogSchema],
      sync: {
        user: app.currentUser,
        flexible: true,
        clientReset: {
          mode: "recoverOrDiscardUnsyncedChanges",
          onBefore: (realm) => {
            // This block could be used for custom recovery, reporting, debugging etc.
          },
          onAfter: (beforeRealm, afterRealm) => {
            // This block could be used for custom recovery, reporting, debugging etc.
          },
          onFallback: (session, path) => {
            // See below "Manual Client Reset Fallback" section for example
          },
        },
      },
    };
    // :snippet-end:
  });

  test.skip("Client reset with recovery fallback", async () => {
    function showUserAConfirmationDialog() {
      return true;
    }
    // :snippet-start: recovery-fallback
    // Must define `realm` at higher scope than `config` so it's accessible
    // from the `onFallback` callback
    // :uncomment-start:
    // let realm;
    // :uncomment-end:

    const config = {
      schema: [DogSchema],
      sync: {
        user: app.currentUser,
        flexible: true,
        clientReset: {
          mode: "recoverOrDiscardUnsyncedChanges", // or "recoverUnsyncedChanges"
          // can also include `onBefore` and `onAfter` callbacks
          onFallback: (_session, path) => {
            try {
              // Prompt user to perform a client reset immediately. If they don't,
              // they won't receive any data from the server until they restart the app
              // and all changes they make will be discarded when the app restarts.
              const didUserConfirmReset = showUserAConfirmationDialog();

              if (didUserConfirmReset) {
                // Close and delete old realm from device
                realm.close();
                Realm.deleteFile(path);

                // Perform client reset
                Realm.App.Sync.initiateClientReset(app, path);

                // Navigate the user back to the main page or reopen the
                // the Realm and reinitialize the current page
              }
            } catch (err) {
              // Reset failed. Notify user that they'll need to
              // update the app
            }
          },
        },
      },
    };

    realm = await Realm.open(config);
    // :snippet-end:
  });

  // skipping because there's no way to manually trigger breaking schema changes
  // client reset from `realm.syncSession._simulateError`. error code `211` should have
  // worked to trigger client reset w breaking schema changes but didn't.
  // this is a known issue being investigated by the JS team.
  // once there is a way to trigger client reset w breaking schema changes from
  // the test, this test will probably have to be refactored slightly to make work.
  test.skip("Discard unsynced changes after breaking schema changes", async () => {
    const DogSchema = {
      name: "Doggo4",
      properties: {
        _id: "objectId",
        name: "string",
        age: "int?",
      },
      primaryKey: "_id",
    };
    const clientResetSuccess = await new Promise((resolve, reject) => {
      // :snippet-start: discard-unsynced-changes-after-destructive-schema-changes
      // Once you have opened your Realm, you will have to keep a reference to it.
      // In the error handler, this reference is called `realm`
      async function handleSyncError(session, syncError) {
        if (syncError.name == "ClientReset") {
          console.log(syncError);
          try {
            console.log("error type is ClientReset....");
            const path = realm.path; // realm.path will not be accessible after realm.close()
            realm.close();
            Realm.App.Sync.initiateClientReset(app, path);

            // Download Realm from the server.
            // Ensure that the backend state is fully downloaded before proceeding,
            // which is the default behavior.
            realm = await Realm.open(config);
            // :remove-start:
            expect(realm.isClosed).toBe(false);
            resolve(true);
            // :remove-end:
            realm.close();
          } catch (err) {
            console.error(err);
            reject(err); // :remove:
          }
        } else {
          // ...handle other error types
          // :remove-start:
          console.log(syncError);
          reject("not a client reset error :(");
          // :remove-end:
        }
      }

      const config = {
        schema: [DogSchema],
        sync: {
          user: app.currentUser,
          flexible: true,
          clientReset: {
            mode: "discardUnsyncedChanges",
            onBefore: (realm) => {
              // NOT used with destructive schema changes
              console.log("Beginning client reset for ", realm.path);
              // :remove-start:
              reject(
                "Test shouldn't touch automatic client reset logic (before)"
              );
              // :remove-end:
            },
            onAfter: (beforeRealm, afterRealm) => {
              // Destructive schema changes do not hit this function.
              // Instead, they go through the error handler.
              console.log("Finished client reset for", beforeRealm.path);
              console.log("New realm path", afterRealm.path);
              // :remove-start:
              reject(
                "Test shouldn't touch automatic client reset logic (after)"
              );
              // :remove-end:
            },
          },
          onError: handleSyncError, // invoked with destructive schema changes
        },
      };
      // :snippet-end:
      Realm.open(config).then((openedRealm) => {
        realm = openedRealm;
        realm.write(() => {
          realm.create("Doggo3", {
            _id: new ObjectId(),
            name: "Maggie",
            age: 13,
          });
        });
        realm.syncSession._simulateError(
          211,
          "Simulate Client Reset",
          "realm::sync::ProtocolError",
          true
        );
      });
    });
    expect(clientResetSuccess).toBe(true);
  });
});

describe("Manual client reset", () => {
  // :snippet-start: track-updates-to-objects
  const DogSchema = {
    name: "Dog",
    properties: {
      name: "string",
      age: "int?",
      lastUpdated: "int",
    },
  };

  // :snippet-end:
  test.skip("Manual client reset - onManual handler", async () => {
    // :snippet-start: manual-client-reset-onmanual
    const config = {
      schema: [DogSchema],
      sync: {
        user: app.currentUser,
        flexible: true,
        clientReset: {
          mode: "manual",
          onManual: (session, path) => {
            // handle manual client reset here
          },
        },
      },
    };
    // :snippet-end:
  });
  test.skip("Manually recover unsynced changes", async () => {
    // :snippet-start: config
    const config = {
      schema: [DogSchema],
      sync: {
        user: app.currentUser,
        partitionValue: "MyPartitionValue",
        clientReset: {
          mode: "manual",
        },
        error: handleSyncError, // callback function defined later
      },
    };
    // :snippet-end:
    const app = new Realm.App({ id: REALM_APP_ID });
    await app.logIn(new Realm.Credentials.anonymous());
    let realm = await Realm.open(config);
    // :snippet-start: last-synced-realm
    const LastSyncedSchema = {
      name: "LastSynced",
      properties: {
        realmTracked: "string",
        timestamp: "int?",
      },
      primaryKey: "realmTracked",
    };
    const lastSyncedConfig = { schema: [LastSyncedSchema] };
    const lastSyncedRealm = await Realm.open(lastSyncedConfig);
    lastSyncedRealm.write(() => {
      lastSyncedRealm.create("LastSynced", {
        realmTracked: "Dog",
      });
    });
    // :snippet-end:
    // :snippet-start: track-sync-session
    // Listens for changes to the Dogs collection
    realm.objects("Dog").addListener(async () => {
      // only update LastSynced if sync session is connected
      // and all local changes are synced
      if (realm.syncSession.isConnected()) {
        await realm.syncSession.uploadAllLocalChanges();
        lastSyncedRealm.write(() => {
          lastSyncedRealm.create("LastSynced", {
            realmTracked: "Dog",
            timestamp: Date.now(),
          });
        });
      }
    });
    // :snippet-end:
    // :snippet-start: handle-sync-error
    async function handleSyncError(_session, error) {
      if (error.name === "ClientReset") {
        const realmPath = realm.path; // realm.path will not be accessible after realm.close()
        realm.close(); // you must close all realms before proceeding

        // pass your realm app instance and realm path to initiateClientReset()
        Realm.App.Sync.initiateClientReset(app, realmPath);

        // Redownload the realm
        realm = await Realm.open(config);
        const oldRealm = await Realm.open(error.config);

        const lastSyncedTime = lastSyncedRealm.objectForPrimaryKey(
          "LastSynced",
          "Dog"
        ).timestamp;
        const unsyncedDogs = oldRealm
          .objects("Dog")
          .filtered(`lastUpdated > ${lastSyncedTime}`);
        // add unsynced dogs to synced realm
        realm.write(() => {
          unsyncedDogs.forEach((dog) => {
            realm.create("Dog", dog, "modified");
          });
        });

        // delete dogs from synced realm that were deleted locally
        const syncedDogs = realm
          .objects("Dog")
          .filtered(`lastUpdated <= ${lastSyncedTime}`);
        realm.write(() => {
          syncedDogs.forEach((dog) => {
            if (!oldRealm.objectForPrimaryKey("Dog", dog._id)) {
              realm.delete(dog);
            }
          });
        });
        // make sure everything syncs and close old realm
        await realm.syncSession.uploadAllLocalChanges();
        oldRealm.close();
      } else {
        console.log(`Received error ${error.message}`);
      }
    }
    // :snippet-end:
  });
});

File: content/realm/examples/cpp/sync/client-reset.cpp
#include <catch2/catch_test_macros.hpp>
#include <cpprealm/sdk.hpp>

static const std::string APP_ID = "cpp-tester-uliix";

namespace realm {

struct FlexibleSync_Dog {
  realm::primary_key<realm::object_id> _id{realm::object_id::generate()};
  std::string name;
  int64_t age;
};
REALM_SCHEMA(FlexibleSync_Dog, _id, name, age)

}  // namespace realm

TEST_CASE("Recover unsynced changes example", "[write]") {
  auto appConfig = realm::App::configuration();
  appConfig.app_id = APP_ID;
  auto app = realm::App(appConfig);

  // :snippet-start: before-after-blocks
  /* You can define blocks to call before and after the client reset occur
     if you need to execute specific logic, such as reporting or debugging. */
  auto beforeReset = [&](realm::db before) {
    /* A block called after a client reset error is detected, but before the
       client recovery process is executed. You could use this block for any
       custom logic, reporting, debugging etc. You have access to the database
       before the client reset occurs in this block. */
  };

  auto afterReset = [&](realm::db device, realm::db server) {
    /* A block called after the client recovery process has executed.
       This block could be used for custom recovery, reporting, debugging etc.
       You have access to the database that is currently on the device - the
       one that can no longer sync - and the new database that has been
       restored from the server. */
  };
  // :snippet-end:

  // :snippet-start: recover-unsynced-changes
  auto user = app.login(realm::App::credentials::anonymous()).get();
  auto syncConfig = user.flexible_sync_configuration();

  // Set the client reset handler with your preferred client reset mode.
  syncConfig.set_client_reset_handler(
      realm::client_reset::recover_unsynced_changes(beforeReset, afterReset));

  auto syncedRealm = realm::db(syncConfig);
  // :snippet-end:

  // Not actually attempting to test a client reset here as the SDK handles
  // that. Just confirming we can write to the server with this config.
  auto updateSubscriptionSuccess =
      syncedRealm.subscriptions()
          .update(
              [](realm::mutable_sync_subscription_set &subs) { subs.clear(); })
          .get();
  CHECK(updateSubscriptionSuccess == true);
  CHECK(syncedRealm.subscriptions().size() == 0);

  updateSubscriptionSuccess =
      syncedRealm.subscriptions()
          .update([](realm::mutable_sync_subscription_set &subs) {
            subs.add<realm::FlexibleSync_Dog>(
                "puppies", [](auto &obj) { return obj.age < 3; });
          })
          .get();
  REQUIRE(updateSubscriptionSuccess == true);
  CHECK(syncedRealm.subscriptions().size() == 1);

  auto syncSession = syncedRealm.get_sync_session();

  auto dog = realm::FlexibleSync_Dog{.name = "Maui", .age = 2};

  syncedRealm.write([&] { syncedRealm.add(std::move(dog)); });

  auto managedDogs = syncedRealm.objects<realm::FlexibleSync_Dog>();
  auto specificDog = managedDogs[0];
  REQUIRE(specificDog.name == "Maui");
  REQUIRE(specificDog.age == static_cast<long long>(2));
  REQUIRE(managedDogs.size() == 1);

  syncedRealm.write([&] { syncedRealm.remove(specificDog); });

  auto managedDogsAfterDelete = syncedRealm.objects<realm::FlexibleSync_Dog>();
  REQUIRE(managedDogsAfterDelete.size() == 0);
  syncSession->wait_for_upload_completion().get();
}

TEST_CASE("Discard unsynced changes example", "[write]") {
  auto appConfig = realm::App::configuration();
  appConfig.app_id = APP_ID;
  auto app = realm::App(appConfig);

  auto beforeReset = [&](realm::db before) {
    // A block called after a client reset error is detected, but before the
  };

  auto afterReset = [&](realm::db device, realm::db server) {
    // A block called after the client recovery process has executed.
  };

  // :snippet-start: discard-unsynced-changes
  auto user = app.login(realm::App::credentials::anonymous()).get();
  auto syncConfig = user.flexible_sync_configuration();

  // Set the client reset handler with your preferred client reset mode.
  syncConfig.set_client_reset_handler(
      realm::client_reset::discard_unsynced_changes(beforeReset, afterReset));

  auto syncedRealm = realm::db(syncConfig);
  // :snippet-end:

  // Not actually attempting to test a client reset here as the SDK handles
  // that. Just confirming we can write to the server with this config.
  auto updateSubscriptionSuccess =
      syncedRealm.subscriptions()
          .update(
              [](realm::mutable_sync_subscription_set &subs) { subs.clear(); })
          .get();
  CHECK(updateSubscriptionSuccess == true);
  CHECK(syncedRealm.subscriptions().size() == 0);

  updateSubscriptionSuccess =
      syncedRealm.subscriptions()
          .update([](realm::mutable_sync_subscription_set &subs) {
            subs.add<realm::FlexibleSync_Dog>(
                "puppies", [](auto &obj) { return obj.age < 3; });
          })
          .get();
  REQUIRE(updateSubscriptionSuccess == true);
  CHECK(syncedRealm.subscriptions().size() == 1);

  auto syncSession = syncedRealm.get_sync_session();

  auto dog = realm::FlexibleSync_Dog{.name = "Maui", .age = 2};

  syncedRealm.write([&] { syncedRealm.add(std::move(dog)); });

  auto managedDogs = syncedRealm.objects<realm::FlexibleSync_Dog>();
  auto specificDog = managedDogs[0];
  REQUIRE(specificDog.name == "Maui");
  REQUIRE(specificDog.age == static_cast<long long>(2));
  REQUIRE(managedDogs.size() == 1);

  syncedRealm.write([&] { syncedRealm.remove(specificDog); });

  auto managedDogsAfterDelete = syncedRealm.objects<realm::FlexibleSync_Dog>();
  REQUIRE(managedDogsAfterDelete.size() == 0);
  syncSession->wait_for_upload_completion().get();
}

TEST_CASE("Manual example", "[write]") {
  auto appConfig = realm::App::configuration();
  appConfig.app_id = APP_ID;
  auto app = realm::App(appConfig);

  // :snippet-start: manual-client-reset
  auto user = app.login(realm::App::credentials::anonymous()).get();
  auto syncConfig = user.flexible_sync_configuration();

  // Set the client reset handler to manual client reset mode.
  syncConfig.set_client_reset_handler(realm::client_reset::manual());

  // Define a Sync error handler for handling the client reset.
  syncConfig.sync_config().set_error_handler(
      [&](realm::sync_session session, realm::sync_error error) {
        if (error.is_client_reset_requested()) {
          /* You might use this for reporting or to instruct the user to delete
             and re-install the app. */
        };
      });

  auto syncedRealm = realm::db(syncConfig);
  // :snippet-end:

  // Not actually attempting to test a client reset here as the SDK handles
  // that. Just confirming we can write to the server with this config.
  auto updateSubscriptionSuccess =
      syncedRealm.subscriptions()
          .update(
              [](realm::mutable_sync_subscription_set &subs) { subs.clear(); })
          .get();
  CHECK(updateSubscriptionSuccess == true);
  CHECK(syncedRealm.subscriptions().size() == 0);

  updateSubscriptionSuccess =
      syncedRealm.subscriptions()
          .update([](realm::mutable_sync_subscription_set &subs) {
            subs.add<realm::FlexibleSync_Dog>(
                "puppies", [](auto &obj) { return obj.age < 3; });
          })
          .get();
  REQUIRE(updateSubscriptionSuccess == true);
  CHECK(syncedRealm.subscriptions().size() == 1);

  auto syncSession = syncedRealm.get_sync_session();

  auto dog = realm::FlexibleSync_Dog{.name = "Maui", .age = 2};

  syncedRealm.write([&] { syncedRealm.add(std::move(dog)); });

  auto managedDogs = syncedRealm.objects<realm::FlexibleSync_Dog>();
  auto specificDog = managedDogs[0];
  REQUIRE(specificDog.name == "Maui");
  REQUIRE(specificDog.age == static_cast<long long>(2));
  REQUIRE(managedDogs.size() == 1);

  syncedRealm.write([&] { syncedRealm.remove(specificDog); });

  auto managedDogsAfterDelete = syncedRealm.objects<realm::FlexibleSync_Dog>();
  REQUIRE(managedDogsAfterDelete.size() == 0);
  syncSession->wait_for_upload_completion().get();
}
